### ISSUE TITLE
Figure out how to persist the appropriate identifiers in the relationships table for PREREQUISITE_FOR and HAS_PREREQUISITE

### ISSUE THREAD

**Issue by jmoseman01** [JohnVCS/fossologyFunTime#22]
Dosocs has the ability to scan packages but the identifier doesn't get generated unless a document gets generated.  I think we might be able scan then generate the document of all the jar files and then finally use the metadata from the graphml file to persist in the relationships table.  One thing that could be problematic is that we may not be able to access these relationships accross seperate document namespaces.  We might be able to design a query to bypass that fact and get the document to print off all the depedencies.

Here's the comments from Thomas T Gurney,  

> Jesse,
> 
> You're correct, with one caveat: Right now DoSOCS doesn't create a package identifier unless you create a document that describes that package. This is because identifiers exist only inside the scope of a single document. There is no concept of a single identifier that is tied to a package and independent of any documents. Hence the concept of a "document namespace" exists.
> 
> In other words, if you just scan a bunch of packages but never create any documents, no identifiers will be created.
> 
> Representing a package-package relationship inside an SPDX document is a tough. Either you'd have to modify the schema to allow a document to describe more than one package (not that difficult but requires writing a bunch of code) or you would have to use external references (basically uncharted territory and I'm not sure how you'd get started on that).
> 
> Hope that helps, and I'll be happy to clarify or elaborate further if needed.
> 
> Tom

**Comment by jmoseman01** [comment]
New email from tom

> Depends on how much time and effort you want to put in to doing things the "right" way vs cobbling together a quick and hacky solution.
> 
> Also it's not 100% clear what your goal is. Assuming you want to specify relationships between packages (package-package) and produce a valid SPDX document with these. In which case...
> 
> The first option you mentioned should work, and it doesn't require you to change the database schema or insert weird rows. However your resulting SPDX document will fail validation because it will have some relationships in it with identifiers that are not defined anywhere else in the document. Still, I think DoSOCS is generating globally unique ID strings (though SPDX spec doesn't require this) so not a huge deal if you don't care if your documents are spec-compliant.
> 
> Second solution doesn't make much sense, as a global document namespace implies a "global document" that describes everything. You would have to write some very ugly queries to extract only the subset of packages you are interested in from this "global document".
> 
> My favorite approach is to use external document references since most of the code for this is already there. External references are basically like an "import" of one document namespace into another so you can have relationships that span across documents. You would just add the row to the relationships table as normal, relating one package to another, but for the document corresponding to the "left" side of the relationship you need to add a row to external_refs pointing to the document namespace of the document on the "right" side of the relationship. You could write a function create_external_refs() and call it from spdxdb.create_document() right at the end. The document template is already written to handle external refs (I think) so they should just show up when added to the database.
> 
> tl;dr: Quick and dirty solution is the first one you suggested. Technically "best" (IMHO) is to use external references.
> 
> Tom

**Comment by jmoseman01** [comment]
> You could write a function create_external_refs() and call it from spdxdb.create_document() right at the end.

The query for external_refs seems to already exist.  This should just show up in the template it seems like.  We will have to create a create_external_refs function within spdxdb.  How we will implement this will still need to be looked into.
